design/one-pager-resource-usages.md

```diff
@@ -0,0 +1,346 @@
+claim kinds and managed resource kinds. A separate controller must reconcile
+each `(resource-usage, managed-resource)` tuple in order to tailor solutions to
+a particular external resource. In the above example the usage controller
+determines the `PostgreSQLInstance` claim is satisfied by a `CloudSQLInstance`
+managed resource and thus creates a `CloudSQLDatabase` and `CloudSQLUser` (both
+of which are modelled in the external CloudSQL API) scoped to the usage.
+
+In the above example:
+
+* `.spec.claimRef` is a reference to the `PostgreSQLInstance` resource claim
+  being consumed.
+* `.spec.consumerRef` is a reference to a consuming resource of an arbitrary
+  kind.
+* `.spec.publish` specifies how the underlying managed resource should be
+  exposed to its consumer. `.spec.publish` and each of its immediate sub-keys
+  are optional.
+* `.spec.publish.service` specifies that an `ExternalName` type `Service` (i.e.
+  a cluster internal DNS CNAME) named `name` should be created pointing to
+  `targetTemplate`, where `targetTemplate` is a Go template expected to render
+  to a target DNS name.
+* `.spec.publish.secret` specifies that a `Secret` named `name` should be
+  created with data built from `dataTemplate`, an object mapping string keys
+  to templated values.
+* `.spec.publish.configMap` specifies that a `ConfigMap` named `name` should be
+  created with data built from `dataTemplate`, an object mapping string keys
+  to templated values.
+* `.status.solutionRefs` represents the array of Kubernetes resources created by
+  the usage controller in order to solve the usage.
+* `.status.publishRefs` represents the array of Kubernetes resources published
+  by the usage controller.
+* `.status.conditions` represents the status of the solving and publishing
+  processes.
+
+Usages are solved before they are published. Each usage controller
+implementation may choose what data is available to the templated fields of its
+usage resource. This allows resource controllers to determine whether usages
+may access any superuser credentials associated with the underlying managed
+resource as appropriate, but puts the onus on controller implementations to
+ensure resource usages are furnished with a consistent data keys regardless of
+the underlying managed resource kind.
+
+## Open Questions
+
+This design is an early draft, and thus has many areas for improvement and
+raises some questions, including:
+
+* Do both resource claims and usages need to exist? If we allowed multiple
+  Kubernetes resources to claim a Crossplane managed resource, or allowed only
+  one usage of a Crossplane managed resource, these two constructs could likely
+  be merged.
+* Does the resource usage concept hold up given the subtle implementation
+  differences between the multiple managed resources that may satisfy a resource
+  claim?
+* Should "sub-managed-resources" such as databases and users be automatically
```

Comment:
Do managed sub-resources warrant another design document? I am interested in the possibility that resource claims may need to be resolved beyond cloud provider APIs.  A MySQL User claim could be resolved through a provider API or through MySQL commands (requiring either a shared MySQL claim or direct access to the secret).  Sub-resources that require parent resource claims may affect the design of how non-root MySQL credentials are bound, for example, and could affect this overall design.

This comment is a bit off-topic, responding only to one the Open Questions.

The overall draft is compelling to me, but I worry about added complexity. I would hope that your other Open Question about merging the constructs could be realized.

Reply:
> Do managed sub-resources warrant another design document?

Quite possibly. They're related to this document and also likely related to any proposals around security connectivity, given a subset of hypothetical security and connectivity resources (e.g. Azure and GCP firewall rules) fall within the set of resources that could potentially be created in order to 'solve' a usage.

> A MySQL User claim could be resolved through a provider API or through MySQL commands (requiring either a shared MySQL claim or direct access to the secret).

Exactly. If we begin to model users and databases it will vary from provider to provider as to whether we can do so simply via API calls (e.g. GCP), or whether we'll need to use SQL client libraries (e.g. AWS, Azure). We might consider whether such resources that are implemented using (for example) SQL client libraries belong in the same binary as resources implemented using cloud provider client libraries.

> Sub-resources that require parent resource claims may affect the design of how non-root MySQL credentials are bound, for example, and could affect this overall design.

I imagine the credentials would be read from the 'parent' managed resource, with a relationship formed between the parent and child managed resources (e.g. `PostgreSQLInstance` and `PostgreSQLUser`) by the resource claim that spawns them (presuming they're spawned by one claim and not two separate ones). A variant of this pattern exists in the Azure storage container controller, which treats the container's associated storage account resource as its 'provider' (instead of a `provider.azure.crossplane.io`) for credential purposes. 

> The overall draft is compelling to me, but I worry about added complexity. 

Agreed. If others find the broad direction compelling we should determine how to reach the goal  iteratively. At the very least I believe we could make the foundational changes (e.g. cluster scoped managed resources, classes, and secrets) first, then introduce some resource usage (or reworked resource claim?) construct in a future change.

Reply:
We discussed offline that the existing `KubernetesCluster` to `KubernetesApplication` / `KubernetesApplicationResource` relationship may serve as a design model for how secrets (cluster credentials) and virtual resources (applications, MySQL users and databases outside of GCP) are treated throughout this project.

@negz you were able to read more into my conjecture.  Would it be beneficial to log your thoughts here?

Reply:
@displague I broadly agree with the thoughts you've captured here. Nothing more to add right now.